docs/language/informal/enhanced-type-promotion.md

```diff
@@ -0,0 +1,540 @@
+There are two places this can happen:
+
+*   Section 16.34 says that an "is-test" expression `v is T` shows that `v` has
+    type T.
+
+*   Section 16.22 propagates facts through logical expressions. It says
+    (roughly) that an and expression like `e1 && e2` shows `v` has type T if
+    either `e1` or `e2` does.
+
+In both cases there is an unstated assumption that the fact is deduced when the
+expression evaluates to true. We extend this to include information that can be
+deduced when an expression evaluates to false. As with the true case, there are
+two places that we can do this.
+
+*   An is-expression `e` of the form `v is! T`, shows that `v` has type T when
+    `e` is false.
+
+*   A logical or expression `e` like `e1 || e2` shows that `v` has type T if
+    either `e1` or `e2` shows that when they are false if `e` itself is false.
+    For example, in `o is! String || itsMonday`, we know `o` must be a String if
+    the entire `||` expression evaluated to false.
+
+## Extending Scope
+
+### "Else" Scope
+
+The specification defines three scopes where promotion can apply:
+
+*   The "then" clause of a conditional expression (16.20):
+
+    ```dart
+    o is String ? oPromotedToStringHere : notHere
+    ```
+
+*   The right operand of a logical `&&` expression (16.22):
+
+    ```dart
+    o is String && oPromotedToStringHere
+    ```
+
+    This is because short-circuiting means the right operand is only evaluated
+    when the left is true.
+
+*   The "then" clause of an if statement (17.5):
+
+    ```dart
+    if (o is String) oPromotedToStringHere;
+    ```
+
+All of these correspond to code that is only reached if a condition expression
+evaluates to true. We extend this to cover facts that are known when an
+expression evaluates to false.
+
+*   In a conditional expression of the form `e1 ? e2 : e3`, we know that if `e1`
+    shows that `v` has type T when `e1` is false, then `v` is known to be of
+    type T in `e3`:
+
+    ```dart
+    o is! String ? notHere : oPromotedToStringHere
+    ```
+
+*   In a logical boolean expression of the form `e1 || e2`, if `e1` shows `v`
+    has type T when `e1` is false, then `v` is known to be of type T in `e2`:
+
+    ```dart
+    o is! String || oPromotedToStringHere
+    ```
+
+    This is because short-circuiting means the right operand is only evaluated
+    when the left is false.
+
+*   In an if statement, if the condition shows that `v` has type T when the
+    condition is false, then `v` is known to be of type T in the else clause.
+
+    ```dart
+    if (o is! String) notHere; else oPromotedToStringHere;
+    ```
+
+### Loop Scopes
+
+Much like if statements, we extend C-style for and while loops to promote inside
+their bodies when the condition is true and the condition expression shows a `v`
+has type T:
+
+```dart
+while (o is String) oPromotedToStringHere;
+
+for (; o is String;;) oPromotedToStringHere;
```

Comment:
If an assignment in the loop body doesn't nullify the promotion then, yes, it seems to me like it should.

@bwilkerson?